src/components/mc/cpu/reduce/mc_cpu_reduce.h

```diff
@@ -162,6 +162,115 @@
         }                                                                      \
     } while (0)
 
+#define OP_BFLOAT16_1(_s1, _s2, _i, _sc, _OP)                                  \
+    _OP(bfloat16tofloat32(&_s1[_i]), bfloat16tofloat32(&_s2[_i]))
+#define OP_BFLOAT16_2(_s1, _s2, _i, _sc, _OP)                                  \
+    _OP((OP_BFLOAT16_1(_s1, _s2, _i, _sc, _OP)),                               \
+        bfloat16tofloat32(&_s2[_i + 1 * _sc]))
+#define OP_BFLOAT16_3(_s1, _s2, _i, _sc, _OP)                                  \
+    _OP((OP_BFLOAT16_2(_s1, _s2, _i, _sc, _OP)),                               \
+        bfloat16tofloat32(&_s2[_i + 2 * _sc]))
+#define OP_BFLOAT16_4(_s1, _s2, _i, _sc, _OP)                                  \
+    _OP((OP_BFLOAT16_3(_s1, _s2, _i, _sc, _OP)),                               \
+        bfloat16tofloat32(&_s2[_i + 3 * _sc]))
+#define OP_BFLOAT16_5(_s1, _s2, _i, _sc, _OP)                                  \
+    _OP((OP_BFLOAT16_4(_s1, _s2, _i, _sc, _OP)),                               \
+        bfloat16tofloat32(&_s2[_i + 4 * _sc]))
+#define OP_BFLOAT16_6(_s1, _s2, _i, _sc, _OP)                                  \
+    _OP((OP_BFLOAT16_5(_s1, _s2, _i, _sc, _OP)),                               \
+        bfloat16tofloat32(&_s2[_i + 5 * _sc]))
+#define OP_BFLOAT16_7(_s1, _s2, _i, _sc, _OP)                                  \
+    _OP((OP_BFLOAT16_6(_s1, _s2, _i, _sc, _OP)),                               \
+        bfloat16tofloat32(&_s2[_i + 6 * _sc]))
+
+#define OP_BFLOAT16_N(_d, _s1, _s2, _sc, _OP, _n)                              \
+    do {                                                                       \
+        size_t _i;                                                             \
+        for (_i = 0; _i < count; _i++) {                                       \
+            float32tobfloat16(OP_BFLOAT16_##_n(_s1, _s2, _i, _sc, _OP),        \
```

Comment:
is this working ? should not there be an assignment d[_i] = float32tobfloat16(...) ?
Did you run gtest/mpitest ? they should have failed. 

Aside: can we do w/o that much code duplication?